---
title: "scnanoseq - BLAZE analysis"
author:
  - name: "Austyn Trull"
  - name: "Lara Ianov, Ph.D."
output:
  html_document:
    code_folding: hide
    css: "style.css"
    toc: true
    toc_float: true
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script performs standard post-secondary analysis on the BLAZE data which has been re-processed with `nf-core/scnanoseq` pipeline. 
This dataset is compared to the published BLAZE data/matrices which we have analyzed in `ubds_analysis_BLAZE_data.Rmd`.

* Reference information: hg38 GENCODE v31 (matching genome version of original BLAZE analysis)

The input are the following files from the pipeline:

```
└── scnanoseq_blaze_data
    ├── ERR9958133.gene_counts.tsv
    ├── ERR9958133.transcript_counts.tsv
    ├── ERR9958134.gene_counts.tsv
    ├── ERR9958134.transcript_counts.tsv
    ├── ERR9958135.gene_counts.tsv
    └── ERR9958135.transcript_counts.tsv
```


```{r}
set.seed(1234)

library(ggplot2)
library(Seurat)
library(dplyr)
library(grid)
library(clustree)
library(rtracklayer)

# add custom functions from UBDS R scripts
lapply(list.files("./scripts/R"), FUN = function(x) source(paste0("./scripts/R/", x)))
```

```{r}
#### CONSTANT DEFS ####

## The prefix to output files
output_prefix <- "./output_scnanoseq_data"

# input files (sample and feature type [gene/transcript])
in_files <- dir("./input/scnanoseq_blaze_data",full.names = TRUE)
in_files

# sample names
sample_names <- dir("./input/scnanoseq_blaze_data",full.names = FALSE)
sample_names <- gsub("\\..*","",sample_names)

# feature
feature_type <- dir("./input/scnanoseq_blaze_data",full.names = FALSE)
feature_type <- gsub("_.*","",feature_type)
feature_type <- gsub(".*\\.","",feature_type)

# clean basenames with feature type
basenames <- paste0(sample_names, "_", feature_type)
basenames
```

# PREPARE INPUT DATA

```{r}
dir.create(file.path(output_prefix), recursive = TRUE)

# Read in the matrices
# At this time, gene and transcript will be part of the same list.
# If needed, feature type can be split at a later time
cell_bc_matrix <- lapply(setNames(in_files,
                                 make.names(basenames)),
               read.table, sep = "\t", header = TRUE, row.names = 1)

# Create the seurat object using RNA Assay
seurat_objs <- mapply(FUN = function(x,y) {
  
  seurat_obj <- CreateSeuratObject(
    counts = x,
    assay = "RNA",
    min.cells = 1,
    min.features = 1,
    project = y)
  
  return(seurat_obj)
  
}, x=cell_bc_matrix, y=names(cell_bc_matrix))
```


# GENERATE QC'S

```{r}
# MITOCHONDRIAL PERCENT

# below can be used if we have gene names and transcript names (transcript names to be carefully checked for pattern)
#seurat_objs <- lapply(X = seurat_objs, FUN = function(x) {
  
  #RNA_MT <- PercentageFeatureSet(object = x, pattern = "^MT-", col.name = "percent.mt", assay = "RNA")
  
#})

#--- Fetch MT features ids ---
# Since currently working with, IDs, they need to be brought in to quantify MT counts

# import gtf
hg38_gtf_gr <- rtracklayer::import("./input/references/gencode.v40.annotation.gtf")

# specify genome
genome(hg38_gtf_gr) <- "GRCh38"

hg38_gtf_gr

### fetch gene names associated with chrM ###

# make vector of all gene names mapped to chrM
hg38_mito_gene_ids <- as.data.frame(hg38_gtf_gr) %>%
  filter(seqnames == "chrM", type == "gene") %>%
  select(seqnames, gene_id, gene_name, gene_type)

# make vector of all transcript names mapped to chrM
hg38_mito_transcript_ids <- as.data.frame(hg38_gtf_gr) %>%
  filter(seqnames == "chrM", type == "transcript") %>%
  select(seqnames, transcript_id, transcript_name, gene_type)

# identify which gene/transcripts are present in the sample(s), and keep only those which are present (prevents error in Seurat)

ids_per_obj <- mapply(FUN = function(x,y) {
  
  if (y == "gene") {
    
    obj_id <- hg38_mito_gene_ids
    
    # which MT genes are in the object
    obj_id$in_data <- hg38_mito_gene_ids$gene_id %in% rownames(x@assays$RNA@counts)
    
    # keep only those which are in the obj
    obj_id <- dplyr::filter(obj_id, in_data==TRUE)
  
  } else {
    
    obj_id <- hg38_mito_transcript_ids
    
    # which MT transcripts are in the object
    obj_id$in_data <- hg38_mito_transcript_ids$transcript_id %in% rownames(x@assays$RNA@counts)
    
    # keep only those which are in the obj
    obj_id <- dplyr::filter(obj_id, in_data==TRUE)
    
  }
  
  return(obj_id)
  
}, x=seurat_objs, y=feature_type, SIMPLIFY = FALSE)

# --- add percentages based on per sample ids ---
seurat_objs <- mapply(FUN = function(x,y) {
  
  RNA_MT <- PercentageFeatureSet(object = x,
                                 features = y[,2], # 2nd col for gene/transcript ids
                                 col.name = "percent.mt",
                                 assay = "RNA")
  
}, x=seurat_objs, y=ids_per_obj)

# brief look at the metadata

lapply(X = seurat_objs, FUN = function(x) {
    
    head(x@meta.data)

})
```

```{r}
# VIOLIN AND FEATURESCATTER, PRE-FILTER QC

mapply(FUN = function(x,y) {
  
  file_name1 <- paste0(output_prefix, "/", names(seurat_objs)[[y]], "_violin_plot_pre_filter.pdf")
  file_name2 <- paste0(output_prefix, "/", names(seurat_objs)[[y]], "_scatter_pre_filter.pdf")
  
  pdf(file = file_name1,
      width = 8,
      height = 6)
  
  plot(VlnPlot(object = x, features = c("nFeature_RNA", "nCount_RNA","percent.mt"), ncol = 3, cols = "#56B4E9", assay = "RNA"))
  
  dev.off()
  
  
  pdf(file = file_name2,
      width = 8,
      height = 6)
  
  plot1 <- FeatureScatter(object = x, feature1 = "nCount_RNA", feature2 = "percent.mt")
  plot2 <- FeatureScatter(object = x, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", cols = "#56B4E9")
  
  plot(plot1 + plot2) 
  
  dev.off()
  
  
}, x = seurat_objs, y = 1:length(seurat_objs))
```

```{r}
# Density plots per sample of nFeatures / nCount (unscaled)

all_plot_nFeatureRNA <- mapply(FUN = function(x) {
  
  p1 <- plotSingleCellDensity(input_obj = x,
                              metadata_variable = "nFeature_RNA",
                              group.by = "orig.ident",
                              scale_x_axis = FALSE) +
  	geom_vline(xintercept = 200)
  
  return(p1)

}, x = seurat_objs, SIMPLIFY = FALSE)

all_plot_nCountRNA <- mapply(FUN = function(x) {
  
  p1 <- plotSingleCellDensity(input_obj = x,
                              metadata_variable = "nCount_RNA",
                              group.by = "orig.ident",
                              scale_x_axis = FALSE)  +
  	geom_vline(xintercept = 300)
  
  return(p1)

}, x = seurat_objs, SIMPLIFY = FALSE)

all_plots_overview_nFeature <- cowplot::plot_grid(plotlist = all_plot_nFeatureRNA)
all_plots_overview_nCount <- cowplot::plot_grid(plotlist = all_plot_nCountRNA)

png(paste0(output_prefix, "/nFeature_un_scaled_RNA_pre_filter.png"), width = 1600, height = 1200)
plot(all_plots_overview_nFeature)
dev.off()

png(paste0(output_prefix, "/nCount_un_scaled_RNA_pre_filter.png"), width = 1600, height = 1200)
plot(all_plots_overview_nCount)
dev.off()
```

Scaled version

```{r, eval=FALSE}
# Density plots per sample of nFeatures / nCount (scaled)

all_plot_nFeatureRNA <- mapply(FUN = function(x) {
  
  p1 <- plotSingleCellDensity(input_obj = x,
                              metadata_variable = "nFeature_RNA",
                              group.by = "orig.ident",
                              scale_x_axis = TRUE) +
  	geom_vline(xintercept = 200)
  
  return(p1)

}, x = seurat_objs, SIMPLIFY = FALSE)

all_plot_nCountRNA <- mapply(FUN = function(x) {
  
  p1 <- plotSingleCellDensity(input_obj = x,
                              metadata_variable = "nCount_RNA",
                              group.by = "orig.ident",
                              scale_x_axis = TRUE) +
  	geom_vline(xintercept = 300)
  
  return(p1)

}, x = seurat_objs, SIMPLIFY = FALSE)

all_plots_overview_nFeature <- cowplot::plot_grid(plotlist = all_plot_nFeatureRNA)
all_plots_overview_nCount <- cowplot::plot_grid(plotlist = all_plot_nCountRNA)

png(paste0(output_prefix, "/nFeature_scaled_RNA_pre_filter.png"), width = 1600, height = 1200)
plot(all_plots_overview_nFeature)
dev.off()

png(paste0(output_prefix, "/nCount_scaled_RNA_pre_filter.png"), width = 1600, height = 1200)
plot(all_plots_overview_nCount)
dev.off()
```


```{r}
# Scatter Plot All 3 Variables

MT_preFil <- purrr::map_dfr(seurat_objs, ~ FetchData(.x, vars = c("orig.ident", "nCount_RNA", "nFeature_RNA",
                                       "percent.mt"))) %>% 
    ggplot(aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) +
  	geom_point() + 
  scale_colour_gradient(low = "grey90", high = "black", limits = c(0,100)) +
  	stat_smooth(method = lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	facet_wrap(~orig.ident, scales = "free") +
    theme_classic()

mt_prefilter_name <- paste0(output_prefix, "/mt_prefilter.png")
png(mt_prefilter_name, width = 1200, height = 800)
plot(MT_preFil)
dev.off()

# Save cell count prior to filtering
cell_count <- purrr::imap_dfr(seurat_objs, ~ FetchData(.x, vars = "orig.ident") %>% 
                                tally(name = "Cells_before_filtering") %>% 
                                mutate(Sample = Project(seurat_objs[[.y]]), .before = `Cells_before_filtering`))

write.csv(cell_count, paste0(output_prefix, "/cell_count_pre_filtering.csv"), row.names = FALSE)
```

# FILTERING

```{r}
# Filtering nFeatures and nCountRNA
#-------NOTE: due to differences in depth between GridION and PromethION, initial cutoff for PromethION is being set a bit higher #----
# This is a general pass at these cutoff for comparisons across the runs.
# Users from the pipeline are advised to careful consider them and change as needed
# Also NOTE: mito % seems a bit high for cortical neuro - a data point to keep in mind from BLAZE data (cutoff at 20% or below)

nFeature_RNA_min <- c(100, 100, 100, 100, 700, 700)

seurat_objs <- mapply(FUN = function(x,y) {
  
  RNA_10X <- subset(x = x, subset  =  nFeature_RNA > y & nFeature_RNA < 999999 & percent.mt < 20)
  
}, x = seurat_objs, y = nFeature_RNA_min)
```

# POST-FILTER QC's

```{r}
# VIOLIN PLOTS

mapply(FUN = function(x,y) {
  
  file_name1 <- paste0(output_prefix, "/", names(seurat_objs)[[y]], "_violin_plot_post_filter.pdf")
  file_name2 <- paste0(output_prefix, "/", names(seurat_objs)[[y]], "_scatter_post_filter.pdf")
  
  pdf(file = file_name1,
      width = 8,
      height = 6)
  
  plot(VlnPlot(object = x, features = c("nFeature_RNA", "nCount_RNA","percent.mt"), ncol = 3, cols = "#56B4E9", assay = "RNA"))
  
  dev.off()
  
  
  pdf(file = file_name2,
      width = 8,
      height = 6)
  
  plot1 <- FeatureScatter(object = x, feature1 = "nCount_RNA", feature2 = "percent.mt")
  plot2 <- FeatureScatter(object = x, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", cols = "#56B4E9")
  
  plot(plot1 + plot2) 
  
  dev.off()
  
  
}, x = seurat_objs, y = 1:length(seurat_objs))
```

Scaled version

```{r, eval=FALSE}
# Density plots per sample of nFeatures / nCount (scaled)

all_plot_nFeatureRNA <- mapply(FUN = function(x) {
  
  p1 <- plotSingleCellDensity(input_obj = x,
                              metadata_variable = "nFeature_RNA",
                              group.by = "orig.ident",
                              scale_x_axis = TRUE)
  
  return(p1)

}, x = seurat_objs, SIMPLIFY = FALSE)

all_plot_nCountRNA <- mapply(FUN = function(x) {
  
  p1 <- plotSingleCellDensity(input_obj = x,
                              metadata_variable = "nCount_RNA",
                              group.by = "orig.ident",
                              scale_x_axis = TRUE)
  
  return(p1)

}, x = seurat_objs, SIMPLIFY = FALSE)

all_plots_overview_nFeature <- cowplot::plot_grid(plotlist = all_plot_nFeatureRNA)
all_plots_overview_nCount <- cowplot::plot_grid(plotlist = all_plot_nCountRNA)

png(paste0(output_prefix, "/nFeature_scaled_RNA_post_filter.png"), width = 1600, height = 1200)
plot(all_plots_overview_nFeature)
dev.off()

png(paste0(output_prefix, "/nCount_scaled_RNA_post_filter.png"), width = 1600, height = 1200)
plot(all_plots_overview_nCount)
dev.off()
```


```{r}
# Scatter Plot All 3 Variables

MT_preFil <- purrr::map_dfr(seurat_objs, ~ FetchData(.x, vars = c("orig.ident", "nCount_RNA", "nFeature_RNA",
                                       "percent.mt"))) %>% 
    ggplot(aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) +
  	geom_point() + 
  scale_colour_gradient(low = "grey90", high = "black", limits = c(0,100)) +
  	stat_smooth(method = lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	facet_wrap(~orig.ident, scales = "free") +
    theme_classic()

mt_prefilter_name <- paste0(output_prefix, "/mt_postfilter.png")
png(mt_prefilter_name, width = 1200, height = 800)
plot(MT_preFil)
dev.off()

# Save cell count prior to filtering
cell_count <- purrr::imap_dfr(seurat_objs, ~ FetchData(.x, vars = "orig.ident") %>% 
                                tally(name = "Cells_after_filtering") %>% 
                                mutate(Sample = Project(seurat_objs[[.y]]), .before = `Cells_after_filtering`))

write.csv(cell_count, paste0(output_prefix, "/cell_count_post_filtering.csv"), row.names = FALSE)
```

# NORMALIZATION, VARIABLE FEATURES, SCALING, PCA
```{r}
seurat_objs <- mapply(FUN = function(x) {
  
  seurat_obj <- NormalizeData(x,
                              assay = "RNA",
                              scale.factor = 10000,
                              normalization.method = "LogNormalize",
                              verbose = TRUE)
  
  seurat_obj <- FindVariableFeatures(seurat_obj,
                                     assay = "RNA",
                                     selection.method = "vst",
                                     nfeatures = 2000)
  
  # add percent.mt as variable to be regressed
  seurat_obj <- ScaleData(seurat_obj,
                          vars.to.regress = "percent.mt")
  
  
  seurat_obj <- RunPCA(seurat_obj, npcs = 50) 
  
  return(seurat_obj)
  
}, x = seurat_objs, SIMPLIFY = FALSE)
```

# ELBOW PLOT

```{r}
mapply(FUN = function(x, y) {
  
  elbow_plot <- paste0(output_prefix, "/", y , "_elbow.png")
  
  png(elbow_plot,
      width = 480,
      height = 480)
  
  plot(ElbowPlot(object = x,
                 ndims = 50,
                 reduction = "pca"))
  
  dev.off()
  
}, x = seurat_objs, y = names(seurat_objs))
```

# DIMENSIONS AND RESOLUTIONS

The loop below will cycle through a set of dimensions (based on elbow plot) and resolutions to aid in a sensible choice of dims and resolution for the integrated object. A temporary object is created for the following process

```{r}
dim_output_dir <- paste0(output_prefix,"/dims_res_explore/")

dir.create(dim_output_dir, recursive = TRUE)

seurat_obj_tmp <- seurat_objs

cluster_res <- seq(from = 0, to = 1, by = 0.1)

dim <- seq(from = 5, to = 15, by = 1)

res <- tidyr::expand_grid(cluster_res, dim)

for (obj in names(seurat_obj_tmp) ) {
  
  out_path <- paste0(dim_output_dir, obj)
  
  dir.create(out_path, recursive = TRUE, showWarnings = FALSE)
  
  for (i in seq(nrow(res))) {
    
    seurat_obj_tmp[[obj]] <- RunUMAP(seurat_obj_tmp[[obj]],
                                     reduction = "pca", 
                                     dims = 1:res[i,]$dim)
    
    seurat_obj_tmp[[obj]] <- FindNeighbors(seurat_obj_tmp[[obj]],
                                           reduction = "pca",
                                           dims = 1:res[i,]$dim)
    
    seurat_obj_tmp[[obj]] <- FindClusters(seurat_obj_tmp[[obj]],
                                          resolution = res[i,]$cluster_res)
    
    file_name <- paste0(out_path,"/",
                        res[i,]$dim,
                        "_res_",
                        res[i,]$cluster_res,
                        ".png")
    
    png(file_name,
        width = 1200,
        height = 800)
    
    plot(
      DimPlot(seurat_obj_tmp[[obj]], label = TRUE) + 
        ggtitle(paste0("Dim_res: ", obj, "_",res[i,]$dim, "_", res[i,]$cluster_res))
    )
    
    dev.off()
    
  }
}

rm(seurat_obj_tmp)
gc()
```

## CLUSTREE

For specific chosen, dimension, generate clustree plot(s)

```{r}
# NOTE: in this case 6 seems fitting for both gene and transcript given low complexity of the dataset.
dim <- 6

tmp_seurat_clustree <- seurat_objs

clustree_res <- seq(from = 0, to = 1, by = 0.1)

res <- tidyr::expand_grid(clustree_res, dim)

for (obj in names(tmp_seurat_clustree) ) {
  
  out_path <- paste0(dim_output_dir, obj)
  
  for (i in seq(nrow(res))) {
    tmp_seurat_clustree[[obj]] <- RunUMAP(tmp_seurat_clustree[[obj]], 
                                          reduction = "pca",
                                          dims = 1:dim)
    
    tmp_seurat_clustree[[obj]] <- FindNeighbors(tmp_seurat_clustree[[obj]],
                                                reduction = "pca",
                                                dims = 1:dim)
    
    tmp_seurat_clustree[[obj]] <- FindClusters(tmp_seurat_clustree[[obj]],
                                               resolution = res[i,]$clustree_res)
  }
  
  # for each object, generate clustree map
  
  clustree_plot <- clustree(tmp_seurat_clustree[[obj]])
  
  pdf(paste0(out_path, "/",
             obj,"_", 
             "clustree.dim_", dim, ".pdf"),
      width = 8,
      height = 12)
  plot(clustree_plot)
  dev.off()
  
}

rm(tmp_seurat_clustree)
gc()
```

## CLUSTER

With appropriate dim and res, cluster the objects:

```{r}
final_dim <- 6
final_res <- 0.6

seurat_objs <- mapply(FUN = function(x) {
  
  seurat_obj <- RunUMAP(obj = x,
                        reduction = "pca",
                        dims = 1:final_dim)
  
  seurat_obj <- FindNeighbors(obj = seurat_obj,
                              reduction = "pca",
                              dims = 1:final_dim)
  
  seurat_obj <- FindClusters(obj = seurat_obj,
                             resolution = final_res)
  
  return(seurat_obj)
  
}, x = seurat_objs)
```

Plot all UMAPs as a single figures across all scnanoseq datasets

```{r}
all_UMAPS <- mapply(FUN = function(x, y) {
  
  p1 <- DimPlot(x,
                reduction = "umap",
                label = FALSE,
                pt.size = 2) + 
    guides(color = guide_legend(override.aes = list(size = 8), ncol=1)) +
    theme(legend.text = element_text(size = 30)) +
    ggtitle(paste0("scnanoseq: ", y))
  
  return(p1)
  
}, x = seurat_objs, y = names(seurat_objs))


png(paste0(output_prefix, "/","All_UMAPS.png"),
    width = 2000,
    height = 1200)

gridExtra::grid.arrange(grobs = all_UMAPS)

dev.off()
```

# INITIAL CLUSTER QC

```{r}
all_UMAPS_Feature <- mapply(FUN = function(x, y) {
  
  p1 <- FeaturePlot(x,
                    features = "nFeature_RNA",
                    reduction = "umap") +
    ggtitle(paste0("scnanoseq BLAZE nFeature: ", y))
  
  return(p1)
  
}, x = seurat_objs, y = names(seurat_objs))


png(paste0(output_prefix, "/","All_nFeature_QC_UMAPs.png"),
    width = 2000,
    height = 1200)

gridExtra::grid.arrange(grobs = all_UMAPS_Feature)

dev.off()


all_UMAPS_Count <- mapply(FUN = function(x, y) {
  
  p1 <- FeaturePlot(x,
                    features = "nCount_RNA",
                    reduction = "umap") +
    ggtitle(paste0("scnanoseq BLAZE nCount: ", y))
  
  return(p1)
  
}, x = seurat_objs, y = names(seurat_objs))


png(paste0(output_prefix, "/","All_nCount_QC_UMAPs.png"),
    width = 2000,
    height = 1200)

gridExtra::grid.arrange(grobs = all_UMAPS_Count)

dev.off()
```

# FindAllMarkers

```{r}
#### CELL MARKERS - FIND ALL MARKERS ####
# FindAllMarkers to recapitulate original analysis

mapply(FUN = function(x, y) {
  
  markers <- FindAllMarkers(x,
                            only.pos = TRUE,
                            min.pct = 0.25,
                            logfc.threshold = 0.25,
                            slot = "data")
  
  #top 20 which pass adjusted p-val
  top_markers <- markers %>%
    dplyr::filter(p_val_adj <= 0.05) %>%
    group_by(cluster) %>%
    slice_max(n = 20, order_by = avg_log2FC)
  
  
  file_name <- paste0(output_prefix, "/", y,"_top_markers.csv")
  
  write.csv(top_markers,
            file_name,
            row.names = FALSE)  
  
  # also saving unfiltered version for reference
  write.csv(markers,
            paste0(output_prefix, "/", y,"_all_markers.csv"),
            row.names = FALSE) 
  
}, x = seurat_objs, y = names(seurat_objs))
```

# BLAZE MARKERS

```{r}
markers <- read.table("input/references/blaze_gene_markers_gencode_v40_full.csv",
                      sep=",",
                      header = TRUE,
                      row.names = NULL)

mapply(FUN = function(x, y, z) {
  
  plot_markers_by_mtx(x,
                      markers,
                      z,
                      paste0(output_prefix, "/markers_", y,"/"),
                      )
  
}, x = seurat_objs, y = names(seurat_objs), z = feature_type)
```

# SAVE PROCESSED OBJECT

Processed object will be saved for cross-comparisons between the analysis (original BLAZE vs BLAZE processed with `scnanoseq`)

```{r}
saveRDS(seurat_objs, file = paste0(output_prefix,"/seurat_scnanoseq_BLAZE.rds"))
```

```{r}
md5 <- paste0("md5sum ", output_prefix, "/seurat_scnanoseq_BLAZE.rds", " > ", output_prefix, "/md5sum.txt")
system(md5)
```
